### PR TITLE
HPCC-30534 Prevent spurious workunit failed states

### PR DESCRIPTION
**Pull request by jakesmith**
The thor agent (managing the instance queue), was spuriously setting workunits to failed, if the Thor instance died, which included when the Thor instance span down when idle (linger). The thoragent manages the instances that jobs target, that in a default configuration (multiLingerJob) will spin up when a wuid needs an instance, and spin down when idle. If whilst and instance was being recycles it threw a k8s exception (e.g. 'Job has reached the specified backoff limit'), it would spuriously cause the original workunit that span up the instance to be marked failed.
The workunit should only be updated at this point, if it is still running.
Normally the workunit workflow instance should manage the final state.

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: ecl/agentexec/agentexec.cpp [hunk 1]
```diff
@@ -337,13 +339,22 @@ class WaitThread : public CInterfaceOf<IPooledThread>
...
-                workunit->commit();
+                // if either a) NOT a thoragent with useChildProcesses=false (default in k8s config) or b) is still in state running
+                if (!sharedK8sJob || (cw->getState() == WUStateRunning))
+                {
+                    // For a shared k8s job, i.e. where this agent is thoragent launching shared (multiJobLinger) k8s jobs
+                    // the job agent should handle the job state.
+                    // In that scenario, this is a fallback that should only come into effect if the job workflow instance has failed to handle the exception
+                    // e.g. because it abruptly disappeared.
+                    Owned<IWorkUnit> workunit = &cw->lock();
+                    workunit->setState(WUStateFailed);
```
Comment (ghalliday) [inline ecl/agentexec/agentexec.cpp:353]: minor: Should probably recheck state now the lock has been obtained.

File: ecl/agentexec/agentexec.cpp [hunk 2]
```diff
@@ -337,13 +339,22 @@ class WaitThread : public CInterfaceOf<IPooledThread>
...
-            if (workunit)
+            Owned<IConstWorkUnit> cw = factory->openWorkUnit(wuid);
+            if (cw)
             {
-                workunit->setState(WUStateFailed);
-                StringBuffer eStr;
-                addExceptionToWorkunit(workunit, SeverityError, "agentexec", exception->errorCode(), exception->errorMessage(eStr).str(), nullptr, 0, 0, 0);
-                workunit->commit();
+                // if either a) NOT a thoragent with useChildProcesses=false (default in k8s config) or b) is still in state running
+                if (!sharedK8sJob || (cw->getState() == WUStateRunning))
```
Comment (ghalliday) [inline ecl/agentexec/agentexec.cpp:346]: Could it possibly be in WUStateBlocked?

**Review by ghalliday — APPROVED** [review]
@jakesmith change looks good.  A question and a comment

File: ecl/agentexec/agentexec.cpp (hunk 1 above)
Comment (jakesmith) [reply ecl/agentexec/agentexec.cpp:353]: good point, it should. I'll change it.

File: ecl/agentexec/agentexec.cpp (hunk 2 above)
Comment (jakesmith) [reply ecl/agentexec/agentexec.cpp:346]: in fact, WUStateWait is also possible. I'll modify.

**Comment by jakesmith** [comment]
@ghalliday - please see new commit.

File: ecl/agentexec/agentexec.cpp [hunk 3]
```diff
@@ -342,18 +342,22 @@ class WaitThread : public CInterfaceOf<IPooledThread>
             Owned<IConstWorkUnit> cw = factory->openWorkUnit(wuid);
             if (cw)
             {
-                // if either a) NOT a thoragent with useChildProcesses=false (default in k8s config) or b) is still in state running
-                if (!sharedK8sJob || (cw->getState() == WUStateRunning))
+                // if either a) NOT a thoragent with useChildProcesses=false (default in k8s config) of b) is still in an executing state
```
Comment (ghalliday) [inline ecl/agentexec/agentexec.cpp:345]: trival accidental change of or to of

**Review by ghalliday — APPROVED** [review]
Looks good.  Please squash, and see trivial comment.

**Comment by jakesmith** [comment]
@ghalliday - fixed comment typo' and squashed.

**Pull request merged**